module/VuFind/src/VuFind/AjaxHandler/GetSearchResults.php

```diff
@@ -0,0 +1,455 @@
+            'target' => 'outer',
+        ],
+        '.js-search-stats' => [
+            'method' => 'renderSearchStats',
+            'target' => 'inner',
+            'attrs' => [
+                'aria-live' => 'polite',
+            ],
+        ],
+        '.js-result-list' => [
+            'method' => 'renderResults',
+            'target' => 'outer',
+        ],
+        'head' => [
+            'method' => 'renderAnalytics',
+            'target' => null,
+        ],
+    ];
+
+    /**
+     * Constructor
+     *
+     * @param SessionSettings  $sessionSettings Session settings
+     * @param ResultsManager   $resultsManager  Results Manager
+     * @param PhpRenderer      $renderer        View renderer
+     * @param RecordLoader     $recordLoader    Record loader
+     * @param ?UserRow         $user            Logged-in user
+     * @param string           $sessionId       Session ID
+     * @param SearchNormalizer $normalizer      Search normalizer
+     * @param SearchTable      $searchTable     Search table
+     * @param array            $config          Main configuration
+     * @param VersionsHelper   $versionsHelper  Versions helper
+     */
+    public function __construct(
+        SessionSettings $sessionSettings,
+        ResultsManager $resultsManager,
+        PhpRenderer $renderer,
+        RecordLoader $recordLoader,
+        ?UserRow $user,
+        string $sessionId,
+        SearchNormalizer $normalizer,
+        Search $searchTable,
+        array $config,
+        VersionsHelper $versionsHelper
+    ) {
+        $this->sessionSettings = $sessionSettings;
+        $this->resultsManager = $resultsManager;
+        $this->renderer = $renderer;
+        $this->recordLoader = $recordLoader;
+        $this->user = $user;
+        $this->sessionId = $sessionId;
+        $this->searchNormalizer = $normalizer;
+        $this->searchTable = $searchTable;
+        $this->config = $config;
+        $this->versionsHelper = $versionsHelper;
+    }
+
+    /**
+     * Handle a request.
+     *
+     * @param Params $params Parameter helper from controller
+     *
+     * @return array [response data, HTTP status code]
+     */
+    public function handleRequest(Params $params)
+    {
+        $results = $this->getSearchResults($params);
+        if (!$results) {
+            return $this->formatResponse(['error' => 'Invalid request'], 400);
+        }
+        $elements = $this->getElements($params, $results);
+        return $this->formatResponse(compact('elements'));
+    }
+
+    /**
+     * Get search results
+     *
+     * @param Params $params Request params
+     *
+     * @return ?Results
+     */
+    protected function getSearchResults(Params $params): ?Results
+    {
+        parse_str($params->fromQuery('querystring', ''), $searchParams);
+        $backend = $params->fromQuery('source', DEFAULT_SEARCH_BACKEND);
+        $searchType = $params->fromQuery('searchType', '');
+        if ('versions' === $searchType) {
```

Comment:
I wonder if it would be beneficial to wrap this in a method like `applyTypeSpecialCases($searchType, & $searchParams): bool` to better encapsulate it (and perhaps encourage more pluggable refactoring in the future, if more special cases arise). The method could return true if search should proceed and false if the caller needs to return null.

Reply:
I eliminated the need for this by passing the history flag to the handler. If someone would mangle the URL and save such search to history, it would still work but display as a normal search, so nothing tragic would happen.